### ISSUE TITLE
Fix `used_underscore_binding`

### ISSUE THREAD

**Issue by mcarton** [rust-lang/rust-clippy#947]
In #944, `used_underscore_binding` was made allow and to ignore `_result` as a ugly hack. This should be fixed.

**Comment by llogiq** [comment]
How? Does a macro check work here?

**Comment by Manishearth** [comment]
We could just have rustc emit `__result` instead

**Comment by bluss** [comment]
using an underscored binding is useful in macros to avoid having the "unused variable" default rustc lint fire.

**Comment by Manishearth** [comment]
rustc uses double underscores for "maybe used idk" bindings in macros, clippy doesn't warn on those

**Comment by oli-obk** [comment]
clippy doesn't warn on used underscore bindings in macros if the binding was declared inside the macro

**Comment by oli-obk** [comment]
I'm currently evaluating if it's possible to insert the hir-lowerings into the expansion-database, so we can treat them just as we treat macros.

**Comment by bluss** [comment]
@oli-obk That must be a recent change

**Comment by oli-obk** [comment]
> @oli-obk That must be a recent change

no, you're right, it isn't checked. I wonder why...

**Comment by oli-obk** [comment]
cc @ThibsG this is the lint we were talking about, but I'm not sure at all whether we do now know that the `_result` from rustc is unused

**Comment by hellow554** [comment]
I came across this today, which is a valid reason to use an underscore variable IMHO:

```rust
let _unused = get_option();
debug_assert!(_unused.is_some());
```

**Comment by kraktus** [comment]
In the above case `debug_assert!(get_option().is_some());` would resolve it though

I triggered it with features flag (pseudo-code):

```rust 

pub fn foo(x: usize) -> usize { internal_foo(x, None) }
#[cfg(feature = "cached")]
pub fn foo_cached(x: usize, cache_key: u64) -> usize { internal_foo(x, Some(cache_key)) } 
fn internal_foo(x: usize, cache_key: Option<u64>) -> usize { ... }
```

This can probably be avoided if I used `internal_foo` as a macro rather than a function, but maybe would be worth adding to the output that if intentional could use `__variable` instead?
